speaker/bgp_controller_test.go

```diff
@@ -222,21 +223,730 @@ func TestBGPSpeaker(t *testing.T) {
+								AggregationLength: 32,
+							},
+						},
+					},
+				},
+			},
+			wantAds: map[string][]*bgp.Advertisement{
+				"1.2.3.4:0": nil,
+			},
+		},
+
+		{
+			desc:     "Add service, not an LB",
+			balancer: "test1",
+			svc: &v1.Service{
+				Spec: v1.ServiceSpec{
+					Type:                  "ClusterIP",
+					ExternalTrafficPolicy: "Cluster",
+				},
+			},
+			eps: k8s.EpsOrSlices{
+				EpVal: &v1.Endpoints{
+					Subsets: []v1.EndpointSubset{
+						{
+							Addresses: []v1.EndpointAddress{
+								{
+									IP:       "2.3.4.5",
+									NodeName: strptr("pandora"),
+								},
+							},
+						},
+					},
+				},
+				Type: k8s.Eps,
+			},
+			wantAds: map[string][]*bgp.Advertisement{
+				"1.2.3.4:0": nil,
+			},
+		},
+
+		{
+			desc:     "Add service, it's an LB!",
+			balancer: "test1",
+			svc: &v1.Service{
+				Spec: v1.ServiceSpec{
+					Type:                  "LoadBalancer",
+					ExternalTrafficPolicy: "Cluster",
+				},
+				Status: statusAssigned("10.20.30.1"),
+			},
+			eps: k8s.EpsOrSlices{
+				EpVal: &v1.Endpoints{
+					Subsets: []v1.EndpointSubset{
+						{
+							Addresses: []v1.EndpointAddress{
+								{
+									IP:       "2.3.4.5",
+									NodeName: strptr("iris"),
+								},
+							},
+						},
+					},
+				},
+				Type: k8s.Eps,
+			},
+			wantAds: map[string][]*bgp.Advertisement{
+				"1.2.3.4:0": {
+					{
+						Prefix: ipnet("10.20.30.1/32"),
+					},
+				},
+			},
+		},
+
+		{
+			desc:     "LB switches to local traffic policy, endpoint isn't on our node",
+			balancer: "test1",
+			svc: &v1.Service{
+				Spec: v1.ServiceSpec{
+					Type:                  "LoadBalancer",
+					ExternalTrafficPolicy: "Local",
+				},
+				Status: statusAssigned("10.20.30.1"),
+			},
+			eps: k8s.EpsOrSlices{
+				EpVal: &v1.Endpoints{
+					Subsets: []v1.EndpointSubset{
+						{
+							Addresses: []v1.EndpointAddress{
+								{
+									IP:       "2.3.4.5",
+									NodeName: strptr("iris"),
+								},
+							},
+						},
+					},
+				},
+				Type: k8s.Eps,
+			},
+			wantAds: map[string][]*bgp.Advertisement{
+				"1.2.3.4:0": nil,
+			},
+		},
+
+		{
+			desc:     "New endpoint, on our node",
+			balancer: "test1",
+			svc: &v1.Service{
+				Spec: v1.ServiceSpec{
+					Type:                  "LoadBalancer",
+					ExternalTrafficPolicy: "Local",
+				},
+				Status: statusAssigned("10.20.30.1"),
+			},
+			eps: k8s.EpsOrSlices{
+				EpVal: &v1.Endpoints{
+					Subsets: []v1.EndpointSubset{
+						{
+							Addresses: []v1.EndpointAddress{
+								{
+									IP:       "2.3.4.5",
+									NodeName: strptr("iris"),
+								},
+								{
+									IP:       "2.3.4.6",
+									NodeName: strptr("pandora"),
+								},
+							},
+						},
+					},
+				},
+				Type: k8s.Eps,
+			},
+			wantAds: map[string][]*bgp.Advertisement{
+				"1.2.3.4:0": {
+					{
+						Prefix: ipnet("10.20.30.1/32"),
+					},
+				},
+			},
+		},
+
+		{
+			desc:     "Endpoint on our node has some unready ports",
```

Comment:
Why does this not advertise anything? The traffic policy is "local" but pandora does have at least one endpoint available for this service.

Reply:
In case of traffic policy = local, having at least one non working endpoint is enough for not advertising the node:
https://github.com/metallb/metallb/blob/0c83b4ba1164ba27a64084a6b0bde12a38c2d24a/speaker/bgp_controller.go#L118

This is the same in the test I cloned.

Reply:
nodeHasHealthyEndpoint() shoud return true if "At least one fully healthy endpoint on this machine.". This is different to your statement above (having at least one non working endpoint is enough for not advertising the node). The host "Pandora" has one ready endpoint so it should announce? I must be missing something as I see the cloned test case is the same and it does pass.

Reply:
Ok, so one Endpoint 2.3.4.6 is not ready on Pandora and the other is - so overall it is not Ready ... that is subtle!

Reply:
Indeed :-)